// firebaseauth/dist/core/endpoints.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const utils = __importStar(require("./utils"));
class EndPoints {
    static post(url, body) {
        return utils.callEndpoint('POST', url, body);
    }
}
EndPoints.urls = (apiKey) => {
    return {
        signInUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword?key=${apiKey}`,
        signUpUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/signupNewUser?key=${apiKey}`,
        sendVerificationEmailUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/getOobConfirmationCode?key=${apiKey}`,
        verifyEmailUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/setAccountInfo?key=${apiKey}`,
        sendPasswordResetEmailUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/getOobConfirmationCode?key=${apiKey}`,
        verifyPasswordResetcodeUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/resetPassword?key=${apiKey}`,
        resetPasswordUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/resetPassword?key=${apiKey}`,
        changePasswordUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/setAccountInfo?key=${apiKey}`,
        accountInfoUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/getAccountInfo?key=${apiKey}`,
        refreshTokenUrl: `https://securetoken.googleapis.com/v1/token?key=${apiKey}`,
        updateAccountInfoUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/setAccountInfo?key=${apiKey}`,
        socialIdentityUrl: `https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyAssertion?key=${apiKey}`
    };
};
exports.default = EndPoints;
